File: app/src/main/java/vn/unlimit/vpngate/dialog/ConnectionUseProtocol.java
package vn.unlimit.vpngate.dialog;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import vn.unlimit.vpngate.R;
import vn.unlimit.vpngate.models.PaidServer;
import vn.unlimit.vpngate.models.VPNGateConnection;

public class ConnectionUseProtocol extends BottomSheetDialogFragment implements View.OnClickListener {
    public interface ClickResult {
        void onResult(boolean useUdp);
    }

    private VPNGateConnection mVpnGateConnection;
    private PaidServer paidServer;
    private Button btnUseTCP;
    private ClickResult clickResult;

    public static ConnectionUseProtocol newInstance(VPNGateConnection vpnGateConnection, ClickResult clickResult) {
        ConnectionUseProtocol connectionUseProtocol = new ConnectionUseProtocol();
        connectionUseProtocol.mVpnGateConnection = vpnGateConnection;
        connectionUseProtocol.clickResult = clickResult;
        return connectionUseProtocol;
    }

    public static ConnectionUseProtocol newInstance(PaidServer paidServer, ClickResult clickResult) {
        ConnectionUseProtocol connectionUseProtocol = new ConnectionUseProtocol();
        connectionUseProtocol.paidServer = paidServer;
        connectionUseProtocol.clickResult = clickResult;
        return connectionUseProtocol;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        BottomSheetDialog dialog =
                new BottomSheetDialog(requireActivity());

        dialog.setOnShowListener(dialog1 -> {
            try {
                BottomSheetDialog d = (BottomSheetDialog) dialog1;

                FrameLayout bottomSheet = d.findViewById(com.google.android.material.R.id.design_bottom_sheet);
                assert bottomSheet != null;
                BottomSheetBehavior.from(bottomSheet).setState(BottomSheetBehavior.STATE_EXPANDED);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        return dialog;
    }

    @SuppressLint({"SetTextI18n", "UseCompatLoadingForDrawables"})
    @Override
    public void setupDialog(@NonNull Dialog dialog, int style) {
        try {
            View contentView = View.inflate(getContext(), R.layout.layout_connect_use_protocol_dialog, null);
            btnUseTCP = contentView.findViewById(R.id.btn_use_tcp);
            btnUseTCP.setOnClickListener(this);
            Button btnUseUDP = contentView.findViewById(R.id.btn_use_udp);
            btnUseUDP.setOnClickListener(this);
            if (mVpnGateConnection != null) {
                btnUseTCP.setText("TCP " + mVpnGateConnection.getTcpPort());
                btnUseUDP.setText("UDP " + mVpnGateConnection.getUdpPort());
            } else if (paidServer != null) {
                btnUseTCP.setBackground(requireContext().getResources().getDrawable(R.drawable.selector_paid_button));
                btnUseTCP.setText("TCP " + paidServer.getTcpPort());
                btnUseUDP.setText("UDP " + paidServer.getUdpPort());
            }
            dialog.setContentView(contentView);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View view) {
        if (clickResult != null) {
            clickResult.onResult(!btnUseTCP.equals(view));
        }
        this.dismiss();
    }
}

File: app/src/main/java/vn/unlimit/vpngate/dialog/ConnectionUseProtocol.kt
package vn.unlimit.vpngate.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.content.res.ResourcesCompat
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import vn.unlimit.vpngate.R
import vn.unlimit.vpngate.databinding.LayoutConnectUseProtocolDialogBinding
import vn.unlimit.vpngate.models.PaidServer
import vn.unlimit.vpngate.models.VPNGateConnection

class ConnectionUseProtocol : BottomSheetDialogFragment(), View.OnClickListener {
    private var mVpnGateConnection: VPNGateConnection? = null
    private var paidServer: PaidServer? = null
    private var clickResult: ClickResult? = null
    private lateinit var binding: LayoutConnectUseProtocolDialogBinding

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog =
            BottomSheetDialog(requireActivity())

        dialog.setOnShowListener { dialog1: DialogInterface ->
            try {
                val d = dialog1 as BottomSheetDialog

                val bottomSheet =
                    checkNotNull(d.findViewById<FrameLayout>(com.google.android.material.R.id.design_bottom_sheet))
                BottomSheetBehavior.from(bottomSheet).setState(BottomSheetBehavior.STATE_EXPANDED)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return dialog
    }

    @SuppressLint("SetTextI18n")
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = LayoutConnectUseProtocolDialogBinding.inflate(layoutInflater)
        binding.btnUseTcp.setOnClickListener(this)
        binding.btnUseUdp.setOnClickListener(this)
        if (mVpnGateConnection != null) {
            binding.btnUseTcp.text = "TCP " + mVpnGateConnection!!.tcpPort
            binding.btnUseUdp.text = "UDP " + mVpnGateConnection!!.udpPort
        } else if (paidServer != null) {
            binding.btnUseTcp.background = ResourcesCompat.getDrawable(requireContext().resources, R.drawable.selector_paid_button, requireActivity().theme)
            binding.btnUseTcp.text = "TCP " + paidServer!!.tcpPort
            binding.btnUseUdp.text = "UDP " + paidServer!!.udpPort
        }
        return binding.root
    }

    override fun onClick(view: View) {
        if (clickResult != null) {
            clickResult!!.onResult(binding.btnUseTcp != view)
        }
        try {
            this.dismiss()
        } catch (e: IllegalStateException) {
            Log.e(TAG, "Got exception when handle on click ", e)
        }
    }

    interface ClickResult {
        fun onResult(useUdp: Boolean)
    }

    companion object {
        const val TAG = "ConnectionUseProtocol"
        fun newInstance(
            vpnGateConnection: VPNGateConnection?,
            clickResult: ClickResult?
        ): ConnectionUseProtocol {
            val connectionUseProtocol = ConnectionUseProtocol()
            connectionUseProtocol.mVpnGateConnection = vpnGateConnection
            connectionUseProtocol.clickResult = clickResult
            return connectionUseProtocol
        }

        fun newInstance(paidServer: PaidServer?, clickResult: ClickResult?): ConnectionUseProtocol {
            val connectionUseProtocol = ConnectionUseProtocol()
            connectionUseProtocol.paidServer = paidServer
            connectionUseProtocol.clickResult = clickResult
            return connectionUseProtocol
        }
    }
}
